src/_pytest/pytester.py

```diff
@@ -1254,80 +1263,80 @@ def run(
             The period in seconds after which to timeout and raise
             :py:class:`Testdir.TimeoutExpired`.
         :param stdin:
-            Optional standard input.  Bytes are being send, closing
+            Optional standard input.  If `bytes`, they are sent, closing
             the pipe, otherwise it is passed through to ``popen``.
             Defaults to ``CLOSE_STDIN``, which translates to using a pipe
             (``subprocess.PIPE``) that gets closed.
-
-        :rtype: RunResult
+        :param encoding:
+            The encoding used to decode the stdout and stderr of the process.
+        :param errors:
+            The error handler to use when decoding the stdout and stderr of the process.
+        :param env:
+            The enviornment variables for the process.
+            The default is to inherit the environment of the current process.
         """
         __tracebackhide__ = True
 
         cmdargs = tuple(
             str(arg) if isinstance(arg, py.path.local) else arg for arg in cmdargs
         )
-        p1 = self.tmpdir.join("stdout")
-        p2 = self.tmpdir.join("stderr")
+        stdout_path = str(self.tmpdir.join("stdout"))
+        stderr_path = str(self.tmpdir.join("stderr"))
+
         print("running:", *cmdargs)
         print("     in:", py.path.local())
-        f1 = open(str(p1), "w", encoding="utf8")
-        f2 = open(str(p2), "w", encoding="utf8")
-        try:
-            now = timing.time()
+
+        stdout = open(stdout_path, "wb")
+        stderr = open(stderr_path, "wb")
+        with stdout, stderr:
+            start = timing.perf_counter()
+
             popen = self.popen(
                 cmdargs,
                 stdin=stdin,
-                stdout=f1,
-                stderr=f2,
+                stdout=stdout,
+                stderr=stderr,
                 close_fds=(sys.platform != "win32"),
             )
             if isinstance(stdin, bytes):
                 popen.stdin.close()
 
-            def handle_timeout() -> None:
-                __tracebackhide__ = True
-
-                timeout_message = (
+            try:
+                ret = popen.wait(timeout)
+            except subprocess.TimeoutExpired:
+                popen.kill()
+                popen.wait()
+                message = (
                     "{seconds} second timeout expired running:"
                     " {command}".format(seconds=timeout, command=cmdargs)
                 )
+                raise self.TimeoutExpired(message)
 
-                popen.kill()
-                popen.wait()
-                raise self.TimeoutExpired(timeout_message)
+            duration = timing.perf_counter() - start
+
+        stdout_text = open(stdout_path, encoding=encoding, errors=errors)
+        stderr_text = open(stderr_path, encoding=encoding, errors=errors)
```

Comment:
there's a potential leak of the first file descriptor if this call were to fail (same above)

Reply:
Fixed, thanks.